core/src/main/java/org/opensearch/sql/calcite/CalciteRelNodeVisitor.java

```diff
@@ -531,8 +533,42 @@ public RelNode visitJoin(Join node, CalcitePlanContext context) {
         node.getJoinCondition()
             .map(c -> rexVisitor.analyzeJoinCondition(c, context))
             .orElse(context.relBuilder.literal(true));
-    context.relBuilder.join(
-        JoinAndLookupUtils.translateJoinType(node.getJoinType()), joinCondition);
+    if (node.getJoinType() == SEMI || node.getJoinType() == ANTI) {
+      // semi and anti join only return left table outputs
+      context.relBuilder.join(
+          JoinAndLookupUtils.translateJoinType(node.getJoinType()), joinCondition);
+    } else {
+      // Join condition could contain duplicated column name, Calcite will rename the duplicated
+      // column name with numeric suffix, e.g. ON t1.id = t2.id, the output contains `id` and `id0`
+      // when a new project add to stack. To avoid `id0`, we will rename the `id0` to `alias.id`
+      // or `tableIdentifier.id`:
+      List<String> leftColumns = context.relBuilder.peek(1).getRowType().getFieldNames();
+      List<String> rightColumns = context.relBuilder.peek().getRowType().getFieldNames();
+      List<String> rightTableName =
```

Comment:
Just to confirm: does these logic below work for multi-join query plan tree recursively?

Reply:
Yes.